### PR TITLE
Provide a way to build CircuitBreakerRule

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Since `CircuitBreakerStrategy` has a few factory methods,
users should implement `CircuitBreakerStrategy` to customize it.
Recently, RetryRuleBuilder is added to Armeria 0.99.5 #2661
In the same way, This PR provides `CircuitBreakerRule` and `CircuitBreakerRuleWithContent`.

Modifications:

- Add `CircuitBreakerRule{WithContent}` and its builder.
- Add `AbstractRuleWithContentBuilder` for deduplicating common code.
- Update documentation.
- Fix a bug where a response fails to duplicate when a single rule subscribes a response multiple times.
- Remove deprecated `RetryStrategy` and `RetryStrategWithContent` in favor of 
    `RetryRule` and `RetryRuleWithContent` 
- Breaking
  - `CircuitBreakerStrategy` and `CircuitBreakerStrategWithContent` has been removed in favor of 
    `CircuitBreakerRule` and `CircuitBreakerRuleWithContent` 

Result:
You can now build complex CircuitBreakerRule conveniently.
```java
CircuitBreakerRule myRule = 
  CircuitBreakerRule.of(
          // A failure if an Exception is raised.
          CircuitBreakerRule.onException(), 
          // Neither a success nor a failure because the request has not been handled by the server.
          CircuitBreakerRule.builder()
                            .onUnprocessed()
                            .thenIgnore(),
          // A failure if the response is 5xx.
          CircuitBreakerRule.onServerErrorStatus(),
          // A success if the response is 2xx.
          CircuitBreakerRule.builder()
                            .onStatusClass(HttpStatusClass.SUCCESS)
                            .thenSuccess(),
          // Neither a success nor a failure. Do not take this response into account.
          CircuitBreakerRule.builder().thenIgnore());
```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        return builder().onException().thenFailure();
+    }
+
+    /**
+     * Returns newly created {@link CircuitBreakerRule} that will report a {@link Response} as a failure,
+     * if an {@link UnprocessedRequestException} which means that the request has not been processed by
+     * the server is raised.
+     */
+    static CircuitBreakerRule onUnprocessed() {
+        return builder().onUnprocessed().thenFailure();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:152]: I think the default should be `thenIgnore` for `onUnprocessed`.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+     * ignores it. If {@link CircuitBreakerDecision#next()} is returned, a next {@link CircuitBreakerRule} will
+     * be evaluated.
+     *
+     * <p>Note that the last {@link CircuitBreakerRule} completes with {@link CircuitBreakerDecision#next()} or
+     * a {@link Response} is not matched with the given {@link CircuitBreakerRule}s, the {@link Response} will
+     * be reported as a success.
+     *
+     * <p>To retrieve the {@link ResponseHeaders}, you can use the specified {@link ClientRequestContext}:
+     *
+     * <p><pre>{@code
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:242]: Let's remove `<p>`

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilder.java [hunk 3]
```diff
@@ -52,16 +51,19 @@
...
+     * Creates a new instance with the specified {@code requestHeadersFilter}.
+     */
     @SuppressWarnings("unchecked")
-    AbstractRetryRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
+    protected AbstractRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
         this.requestHeadersFilter = (Predicate<RequestHeaders>) requestHeadersFilter;
     }
 
     /**
      * Adds the specified {@code responseHeadersFilter} for a {@link RetryRule} which will retry
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilder.java:63]: This class is not just for `RetryRule` anymore.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public static BiFunction<? super ClientRequestContext, ? super Throwable, Boolean>
+    buildFilter(AbstractRuleBuilder builder) {
+        final Predicate<RequestHeaders> requestHeadersFilter = builder.requestHeadersFilter();
+        final Predicate<ResponseHeaders> responseHeadersFilter = builder.responseHeadersFilter();
+        final Predicate<Throwable> exceptionFilter = builder.exceptionFilter();
+        return (ctx, cause) -> {
+            if (ctx.log().isAvailable(RequestLogProperty.REQUEST_HEADERS)) {
+                final RequestHeaders requestHeaders = ctx.log().partial().requestHeaders();
+                if (!requestHeadersFilter.test(requestHeaders)) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:43]: `requestHeadersFilter != null`?

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                    return false;
+                }
+            }
+
+            if (requestHeadersFilter == null && responseHeadersFilter == null) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:48]: Or are these always non null?

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleWithContentBuilder.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    @Nullable
+    private Function<? super T, ? extends CompletionStage<Boolean>> responseFilter;
+
+    protected AbstractRuleWithContentBuilder(
+            Predicate<? super RequestHeaders> requestHeadersFilter) {
+        super(requestHeadersFilter);
+    }
+
+    /**
+     * Adds the specified {@code responseFilter} for a {@link AbstractRuleWithContentBuilder}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleWithContentBuilder.java:51]: a -> an

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java (hunk 4 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:43]: `requestHeadersFilter` is not null  because it always contains HTTP methods filter.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:44]: Then we can remove the line 48, right?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+package com.linecorp.armeria.client.circuitbreaker;
+
+import com.linecorp.armeria.common.Response;
+
+/**
+ * A {@link CircuitBreakerDecision} that determines whether a {@link CircuitBreakerRule} reports a response as
+ * a {@link #success()}, {@link #failure()}, {@link #ignore()}s it.
+ * {@link #next()} skips a {@link CircuitBreakerRule} and will lookup next {@link CircuitBreakerRule}s.
+ */
+public final class CircuitBreakerDecision {
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java:26]: Yeah, I intended to provide an API design similar to `RetryDecision`.
However, enum is a good choice too.

[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java:26]: I think `RetryDecision` is not enum because of this:
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/retry/RetryDecision.java#L35
Which is that a user might create a new instance with a different `Backoff`.
But for `CircuitBreakerDecision`, we don't have such use cases.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:152]: Ignoring on unprocessed seems default behavior.
However, other factory methods build a rule with `thenFailure()`.
Probably better to remove this factory to reduce confusion?

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilder.java (hunk 3 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilder.java]: Yeap, I cleaned up Javadoc.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java (hunk 5 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:49]: Oops... It is a bug. It should be:
```suggestion
            if (exceptionFilter == null && responseHeadersFilter == null) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:152]: +1 for removing. 😄

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java [hunk 8]
```diff
@@ -145,29 +257,31 @@ protected HttpResponse doExecute(ClientRequestContext ctx, HttpRequest req, Circ
         try {
             response = delegate().execute(ctx, req);
         } catch (Throwable cause) {
-            if (needsContentInStrategy) {
-                reportSuccessOrFailure(circuitBreaker, strategyWithContent().shouldReportAsSuccess(
-                        ctx, HttpResponse.ofFailure(cause)));
+            if (needsContentInRule) {
+                reportSuccessOrFailure(circuitBreaker, ruleWithContent().shouldReportAsSuccess(
+                        ctx, HttpResponse.ofFailure(cause), cause));
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:262]: Yes. the `content` should not be null yet. If a user set `CircuitBreakerStrategyWithContent`, the content is passed to it via `CircuitBreakerRuleWithContent`
https://github.com/line/armeria/blob/27958cc7219b9c132ba5efab6154fdf09644671b/core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleUtil.java#L68

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java [hunk 9]
```diff
@@ -145,29 +257,31 @@ protected HttpResponse doExecute(ClientRequestContext ctx, HttpRequest req, Circ
...
-
-        ctx.log().whenAvailable(RequestLogProperty.RESPONSE_HEADERS).thenAccept(log -> {
-            final Throwable cause =
-                    log.isAvailable(RequestLogProperty.RESPONSE_CAUSE) ? log.responseCause() : null;
-            reportSuccessOrFailure(circuitBreaker, strategy().shouldReportAsSuccess(ctx, cause));
-        });
-        return response;
+        final CompletableFuture<HttpResponse> responseFuture =
+                ctx.log().whenAvailable(RequestLogProperty.RESPONSE_HEADERS).thenApply(log -> {
+                    if (needsContentInRule) {
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:271]: Good point. But cannot pass null yet(https://github.com/line/armeria/pull/2726#discussion_r427026301) 
Let me pass `shouldReportAsSuccess(ctx, HttpResponse.ofFailure(cause), cause)` without duplicator.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:262]: By the way, I realized null content is passed to `RetryStrategyWithContent`. 😱
https://github.com/line/armeria/blob/ac3440be28a34d436e4382257d46bbf2cbdbca8a/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java#L329-L336

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 10]
```diff
@@ -52,16 +52,18 @@
...
     @SuppressWarnings("unchecked")
-    AbstractRetryRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
+    protected AbstractRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
         this.requestHeadersFilter = (Predicate<RequestHeaders>) requestHeadersFilter;
     }
 
     /**
-     * Adds the specified {@code responseHeadersFilter} for a {@link RetryRule} which will retry
-     * if the {@code responseHeadersFilter} returns {@code true}.
+     * Adds the specified {@code responseHeadersFilter} for a {@link RetryRule} or {@link CircuitBreakerRule}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:64]: How about just removing `for a ... or ...`?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 11]
```diff
@@ -29,21 +28,22 @@
...
+import com.linecorp.armeria.client.retry.RetryRuleWithContent;
 import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.common.HttpStatusClass;
 import com.linecorp.armeria.common.RequestHeaders;
 import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.util.UnstableApi;
 
 /**
- * An abstract builder class which creates a {@link RetryRule} or a {@link RetryRuleWithContent}.
+ * An abstract builder class which creates a {@link RetryRule}, a {@link RetryRuleWithContent},
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:42]: A skeletal builder implementation for ... and ...? (like `AbstractRuleWithContentBuilder`)

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    @Nullable
+    private Function<? super T, ? extends CompletionStage<Boolean>> responseFilter;
+
+    protected AbstractRuleWithContentBuilder(
+            Predicate<? super RequestHeaders> requestHeadersFilter) {
+        super(requestHeadersFilter);
+    }
+
+    /**
+     * Adds the specified {@code responseFilter} for a {@link AbstractRuleWithContentBuilder}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java]: Ditto - Could remove `for ...`

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+     * >     }
+     *
+     * >     ResponseHeaders responseHeaders = ctx.log().responseHeaders();
+     * >     if (responseHeaders.status().codeClass() == HttpStatusClass.SERVER_ERROR) {
+     * >         return CompletableFuture.completedFuture(CircuitBreakerDecision.failure());
+     * >     }
+     * >     ...
+     * >     return CompletableFuture.completedFuture(CircuitBreakerDecision.success())
+     * > }
+     * }</pre>
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:256]: .. and close `<p>` here

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleBuilder.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+     * {@link #thenIgnore()} if an {@link Exception} is raised.
+     */
+    @Override
+    public CircuitBreakerRuleBuilder onException() {
+        return (CircuitBreakerRuleBuilder) super.onException();
+    }
+
+    /**
+     * Makes a {@link CircuitBreakerRule} report a {@link Response} as a success, failure or ignore it
+     * according to the build methods - {@link #thenSuccess()}, {@link #thenFailure()} and
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleBuilder.java:207]: Could you rephrase `as ....` part? Not correct grammar.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleBuilder.java [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+     * {@link #thenIgnore()}, if an {@link UnprocessedRequestException} which means that the request has not
+     * been processed by the server is raised.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleBuilder.java:209]: Insert a comma before `which` and `is` for readability.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleUtil.java [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+final class CircuitBreakerRuleUtil {
+
+    public static final CompletableFuture<CircuitBreakerDecision> SUCCESS_DECISION =
+            CompletableFuture.completedFuture(CircuitBreakerDecision.success());
+    public static final CompletableFuture<CircuitBreakerDecision> FAILURE_DECISION =
+            CompletableFuture.completedFuture(CircuitBreakerDecision.failure());
+    public static final CompletableFuture<CircuitBreakerDecision> IGNORE_DECISION =
+            CompletableFuture.completedFuture(CircuitBreakerDecision.ignore());
+    public static final CompletableFuture<CircuitBreakerDecision> NEXT_DECISION =
+            CompletableFuture.completedFuture(CircuitBreakerDecision.next());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleUtil.java:38]: `public` could be removed?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContent.java [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        checkArgument(circuitBreakerRules.length > 0, "circuitBreakerRules can't be empty.");
+        if (circuitBreakerRules.length == 1) {
+            return circuitBreakerRules[0];
+        }
+        return of(ImmutableList.copyOf(circuitBreakerRules));
+    }
+
+    /**
+     * Returns a {@link CircuitBreakerRuleWithContent} that combines all the
+     * {@link CircuitBreakerRuleWithContent} of the {@code circuitBreakerRules}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContent.java:110]: `combines the specified {@link CircuitBreakerRuleWithContent}s`?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContent.java [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+     * {@code requestHeadersFilter}.
+     */
+    static <T extends Response> CircuitBreakerRuleWithContentBuilder<T> builder(
+            Predicate<? super RequestHeaders> requestHeadersFilter) {
+        requireNonNull(requestHeadersFilter, "requestHeadersFilter");
+        return new CircuitBreakerRuleWithContentBuilder<>(requestHeadersFilter);
+    }
+
+    /**
+     * Returns a {@link CircuitBreakerRuleWithContent} that combines the specified {@code circuitBreakerRules}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContent.java:95]: `combines the specified {@link CircuitBreakerRuleWithContent}s`?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+import com.linecorp.armeria.client.AbstractRuleWithContentBuilder;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.ResponseHeaders;
+
+/**
+ * A builder that creates {@link CircuitBreakerRuleWithContent}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java:34]: creates a

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+ */
+public class CircuitBreakerRuleWithContentBuilder<T extends Response>
+        extends AbstractRuleWithContentBuilder<T> {
+
+    CircuitBreakerRuleWithContentBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
+        super(requestHeadersFilter);
+    }
+
+    /**
+     * Returns newly created {@link CircuitBreakerRuleWithContent} that determines whether a {@link Response}
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java:45]: Many missing a/an

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+     * should be reported as a success.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java:46]: `determines a {@link Response} as a success when the rule matches.`

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java:26]: I thought users might not be interested in this is enum or not.
The important thing is how to use those APIs. Some API they should use `next()` and others are not.
```java
RetryDecision.next() 
// vs
CircuitBreakerDecision.NEXT
// vs
CircuitBreakerDecision.next()
```
By the way, I'm not strong here. Let me migrate this to enum.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java (hunk 4 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:44]: It is a fast path. The bug was fixed. https://github.com/line/armeria/pull/2726#discussion_r427024141

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java:26]: I'm not strong on this neither. 😆 If you think, the current implementation is better in terms of providing a similar API with `RetryDecision` then it's good as it is.

[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerDecision.java:27]: Then let me keep the current implementation.😉

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 22]
```diff
@@ -29,21 +28,19 @@
...
+import com.linecorp.armeria.client.retry.RetryRuleWithContent;
 import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.common.HttpStatusClass;
 import com.linecorp.armeria.common.RequestHeaders;
 import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.util.UnstableApi;
 
 /**
- * An abstract builder class which creates a {@link RetryRule} or a {@link RetryRuleWithContent}.
+ * A skeletal builder implementation for {@link RetryRule} and {@link RetryRuleWithContent}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:40]: I realized this class can be used for building `CircuitBreakerRule(WithContent)`.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+import com.linecorp.armeria.common.ResponseHeaders;
+
+/**
+ * Determines whether a {@link Response} should be reported as a success or failure to a
+ * {@link CircuitBreaker}. If you need to determine whether the request was successful by looking into the
+ * {@link Response} content, use {@link CircuitBreakerRuleWithContent}.
+ *
+ * <p>Note that the last {@link CircuitBreakerRule} completes with {@link CircuitBreakerDecision#next()} or
+ * a {@link Response} is not matched with the {@link CircuitBreakerRule}s, the {@link Response} will be
+ * reported as a success.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRule.java:48]: Grammar?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public CircuitBreakerRuleWithContentBuilder<T> onStatus(Predicate<? super HttpStatus> statusFilter) {
+        return (CircuitBreakerRuleWithContentBuilder<T>) super.onStatus(statusFilter);
+    }
+
+    /**
+     * Adds the specified exception type for a {@link CircuitBreakerRuleWithContent}.
+     * If an {@link Exception} is raised and that is instance of the specified {@code exception},
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRuleWithContentBuilder.java:192]: and that is -> and it is an

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 25]
```diff
@@ -160,32 +151,39 @@ public AbstractRetryRuleBuilder onException(Predicate<? super Throwable> excepti
...
-    public AbstractRetryRuleBuilder onUnprocessed() {
+    public AbstractRuleBuilder onUnprocessed() {
         return onException(UnprocessedRequestException.class);
     }
 
-    Predicate<RequestHeaders> requestHeadersFilter() {
+    /**
+     * Returns the {@link Predicate} of a {@link RequestHeaders}.
+     */
+    public final Predicate<RequestHeaders> requestHeadersFilter() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:170]: protected?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 26]
```diff
@@ -160,32 +151,39 @@ public AbstractRetryRuleBuilder onException(Predicate<? super Throwable> excepti
...
         return requestHeadersFilter;
     }
 
+    /**
+     * Returns the {@link Predicate} of a {@link ResponseHeaders}.
+     */
     @Nullable
-    Predicate<ResponseHeaders> responseHeadersFilter() {
+    public final Predicate<ResponseHeaders> responseHeadersFilter() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:178]: protected?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 27]
```diff
@@ -160,32 +151,39 @@ public AbstractRetryRuleBuilder onException(Predicate<? super Throwable> excepti
...
         return responseHeadersFilter;
     }
 
+    /**
+     * Returns the {@link Predicate} of an {@link Exception}.
+     */
     @Nullable
-    Predicate<Throwable> exceptionFilter() {
+    public final Predicate<Throwable> exceptionFilter() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:186]: protected?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+            };
+        }
+        return this;
+    }
+
+    /**
+     * Returns the {@code responseFilter}.
+     */
+    @Nullable
+    protected final Function<? super T, ? extends CompletionStage<Boolean>> responseFilter() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java:97]: protected?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java (hunk 25 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:170]: These getter are used in `internal.client.AbstractRuleBuilderUtil`. 😅

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:170]: Let me refactor it.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java [hunk 29]
```diff
@@ -75,35 +86,54 @@ public RetryingClientBuilder useRetryAfter(boolean useRetryAfter) {
...
+     *
+     * @param contentPreviewLength the content length to preview. {@code 0} does not disable the length limit.
      *
      * @throws IllegalStateException if this builder is created with a {@link RetryRule} rather than
      *                               {@link RetryRuleWithContent}
-     * @throws IllegalArgumentException if the specified {@code contentPreviewLength} is equal to or
+     * @throws IllegalArgumentException if the specified {@code maxContentLength} is equal to or
      *                                  less than {@code 0}
+     *
+     * @deprecated Use {@link #maxContentLength(int)}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java:123]: I think we should guide to use `builder(...)` not the deprecated one.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java [hunk 30]
```diff
@@ -75,35 +86,54 @@ public RetryingClientBuilder useRetryAfter(boolean useRetryAfter) {
...
+     * @throws IllegalStateException if this builder is created with a {@link RetryRule} rather than
+     *                               {@link RetryRuleWithContent}
+     * @throws IllegalArgumentException if the specified {@code maxContentLength} is equal to or
+     *                                  less than {@code 0}
      *
-     * @return {@link RetryingClientBuilder} to support method chaining
+     * @deprecated Use {@link RetryingClient#builder(RetryRuleWithContent, int)}.
+     */
+    @Deprecated
+    public RetryingClientBuilder maxContentLength(int maxContentLength) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java:99]: Then, we can remove this.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+        return (ctx, cause) -> {
+            if (ctx.log().isAvailable(RequestLogProperty.REQUEST_HEADERS)) {
+                final RequestHeaders requestHeaders = ctx.log().partial().requestHeaders();
+                if (!requestHeadersFilter.test(requestHeaders)) {
+                    return false;
+                }
+            }
+
+            // Safe to return true since no filters are set
+            if (exceptionFilter == null && responseHeadersFilter == null && !hasResponseFilter) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:49]: We can just check `!hasResponseFilter` or add `@Nullable` to parmaters.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+                return true;
+            }
+
+            if (cause != null && exceptionFilter != null && exceptionFilter.test(Exceptions.peel(cause))) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:53]: ditto for `exceptionFilter`

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+                return true;
+            }
+
+            if (ctx.log().isAvailable(RequestLogProperty.RESPONSE_HEADERS)) {
+                final ResponseHeaders responseHeaders = ctx.log().partial().responseHeaders();
+                if (responseHeadersFilter != null && responseHeadersFilter.test(responseHeaders)) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:59]: ditto for `responseHeadersFilter`

File: core/src/main/java/com/linecorp/armeria/common/HttpMessageAggregator.java [hunk 34]
```diff
@@ -62,6 +62,10 @@ public final void onError(Throwable t) {
 
     @Override
     public final void onComplete() {
+        if (future.isDone()) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpMessageAggregator.java:65]: If `onComplete()` called multiple times, this method throws null pointer exception at 92 line.
contentList is 
```java
// All items in contentList was set to null in the previous `onComplete()` call
final HttpData data = contentList.set(i, null); 
// data is null
final int dataLength = data.length();
```

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java (hunk 29 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java:123]: Oops, I should remove this method.
```java
RetryClient.builder(RetryRuleWithContent); // Will use default max content length value
RetryClient.builder(RetryRuleWithContent, maxContentLength); // Will use the specified value

// Deprecated
RetryClient.builder(RetryRuleWithContent).contentPreviewLength(maxContentLength);
```
